### DISCUSSION TITLE
Search for messages by specific filter criteria in Client API returns java.lang.NullPointerException

### DISCUSSION THREAD

**Discussion by AvishkaLiyanaarachchi** [nextgenhealthcare/connect#4638] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/49421667/126109974-8bb7ae44-f40e-4cb6-9159-1a177665dd14.png)

I was able to save ADT message PID segment data to db as above image and now I am trying to query specific column details using 
channel id as bellow 

![image](https://user-images.githubusercontent.com/49421667/126110111-f7a28b98-b75e-448d-82f0-73822bb9a9f6.png)

but I am getting following error even when searching using channel id and additional values 

![image](https://user-images.githubusercontent.com/49421667/126110232-0f1f0e42-a039-4710-b7f0-ebd1109fe9c4.png)

Would like to get help on this pls.

**Comment by jonbartels** [comment]
What MC version? The community here would need to know that to know what code correlates to the lines given in the stack trace.

Can you please also post the stack trace as text instead of an image? Please use code tags. This helps read and search the trace for the root cause.

This sounds similar to a problem discussed in the Mirth Connect Slack channel. The problem at the time was that the MC API required certain fields but just threw a generic exception instead of a missing fields error when they were not given.

**Reply by AvishkaLiyanaarachchi** [reply]
MC version = 3.11.0

only required field is channel id for search message by specific filter criteria

**Stack trace**
`<java.lang.NullPointerException>
  <stackTrace>
    <trace>com.mirth.connect.server.controllers.DonkeyMessageController.getMessages(DonkeyMessageController.java:197)</trace>
    <trace>com.mirth.connect.server.api.servlets.MessageServlet.getMessages(MessageServlet.java:169)</trace>
    <trace>sun.reflect.GeneratedMethodAccessor246.invoke(Unknown Source)</trace>
    <trace>sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)</trace>
    <trace>java.lang.reflect.Method.invoke(Method.java:498)</trace>
    <trace>com.mirth.connect.server.api.providers.MirthResourceInvocationHandlerProvider$1.invoke(MirthResourceInvocationHandlerProvider.java:219)</trace>
    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)</trace>
    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)</trace>
    <trace>org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)</trace>
    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)</trace>
    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)</trace>
    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)</trace>
    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)</trace>
    <trace>org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)</trace>
    <trace>org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)</trace>
    <trace>org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)</trace>
    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:315)</trace>
    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:297)</trace>
    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:267)</trace>
    <trace>org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)</trace>
    <trace>org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)</trace>
    <trace>org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)</trace>
    <trace>org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:471)</trace>
    <trace>org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:425)</trace>
    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:383)</trace>
    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:336)</trace>
    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:223)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:760)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)</trace>
    <trace>com.mirth.connect.server.api.providers.StrictTransportSecurityFilter.doFilter(StrictTransportSecurityFilter.java:33)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)</trace>
    <trace>com.mirth.connect.server.MethodFilter.doFilter(MethodFilter.java:37)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)</trace>
    <trace>com.mirth.connect.server.api.providers.ClickjackingFilter.doFilter(ClickjackingFilter.java:44)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)</trace>
    <trace>com.mirth.connect.server.api.providers.ApiOriginFilter.doFilter(ApiOriginFilter.java:71)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)</trace>
    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)</trace>
    <trace>org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1589)</trace>
    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)</trace>
    <trace>org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1296)</trace>
    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)</trace>
    <trace>org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)</trace>
    <trace>org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1559)</trace>
    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)</trace>
    <trace>org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1211)</trace>
    <trace>org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)</trace>
    <trace>org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)</trace>
    <trace>org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)</trace>
    <trace>org.eclipse.jetty.server.Server.handle(Server.java:500)</trace>
    <trace>org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:386)</trace>
    <trace>org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:560)</trace>
    <trace>org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:378)</trace>
    <trace>org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268)</trace>
    <trace>org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)</trace>
    <trace>org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)</trace>
    <trace>org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:426)</trace>
    <trace>org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:320)</trace>
    <trace>org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:158)</trace>
    <trace>org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)</trace>
    <trace>org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)</trace>
    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)</trace>
    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)</trace>
    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)</trace>
    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)</trace>
    <trace>org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367)</trace>
    <trace>org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)</trace>
    <trace>org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:914)</trace>
    <trace>java.lang.Thread.run(Thread.java:745)</trace>
  </stackTrace>
  <suppressedExceptions class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
    <c class="list"/>
    <list/>
  </suppressedExceptions>
</java.lang.NullPointerException>`

**Reply by pacmano1** [reply]
The documentation is not correct.   Someone over at the forums at the same problem the other day.  It has been discussed on Slack as well but the history is gone.

**Reply by jonbartels** [reply]
@pacmano1  I'm trying to find the history on this issue. I remember the same Slack conversation but I cannot find a solid reference for it.

A forum user posted nearly the same error at https://forums.mirthproject.io/forum/mirth-connect/support/17382-_search-endpoint-xml-schema but no resolution. 

I think when this came up in Slack the root case was a field that was required for search but was not documented and not validated. I remember the problem because some block of code had a call pattern which hid the actual null value or line numbers. 

@AvishkaLiyanaarachchi  - Can you look at your server logs for MC in the file `mirth.log` and post any stack traces shown there?

**Reply by AvishkaLiyanaarachchi** [reply]
When execute the get request in swagger not updating or creating a log file and coming same error. 

mirth.log

I trying to send it as a stack trace but coming error as file is too long so i upload it as a link.

[aaaa.txt](https://github.com/nextgenhealthcare/connect/files/6852883/aaaa.txt)

**Comment by jonbartels** [comment]
Found a solution in Slack history - https://mirthconnect.slack.com/archives/C02SW0K4D/p1622834343161500

It looks like adding the "offset" parameter will solve this issue.

**Comment by AdrianRossouw** [comment, marked as answer]
I did some testing here, and it turns out that 'limit', 'offset' and 'includeContent' are all required : 
<img width="703" alt="Screenshot 2022-02-01 at 16 29 57" src="https://user-images.githubusercontent.com/115108/152009239-064df499-d54a-4317-ba1c-b0dfa4591fa2.png">

**Comment by jonbartels** [comment]
https://github.com/nextgenhealthcare/connect/issues/5203 may have fixed this issue
